src/client-agent/start_agent.c

```diff
@@ -166,134 +169,193 @@ void start_agent(int is_startup)
+        }
+        
+        if (recv_b > 0) {
+            /* Id of zero -- only one key allowed */
+            if (ReadSecMSG(&keys, buffer, cleartext, 0, recv_b - 1, &msg_length, agt->server[agt->rip_id].rip, &tmp_msg) != KS_VALID) {
+                mwarn(MSG_ERROR, agt->server[agt->rip_id].rip);
             } else {
-                recv_b = recv(agt->sock, buffer, OS_MAXSTR, MSG_DONTWAIT);
+                /* Check for commands */
+                if (IsValidHeader(tmp_msg)) {
+                    /* If it is an ack reply */
+                    if (strcmp(tmp_msg, HC_ACK) == 0) {
+                        available_server = time(0);
+
+                        minfo(AG_CONNECTED, agt->server[agt->rip_id].rip,
+                                agt->server[agt->rip_id].port, agt->server[agt->rip_id].protocol == IPPROTO_UDP ? "udp" : "tcp");
+
+                        if (is_startup) {
+                            /* Send log message about start up */
+                            snprintf(msg, OS_MAXSTR, OS_AG_STARTED,
+                                    keys.keyentries[0]->name,
+                                    keys.keyentries[0]->ip->ip);
+                            snprintf(fmsg, OS_MAXSTR, "%c:%s:%s", LOCALFILE_MQ,
+                                    "ossec", msg);
+                            send_msg(fmsg, -1);
+                        }
+                        return;
+                    }
+                }
             }
+        }
+
+        /* If thre is a next server, try it */
+        if (agt->server[agt->rip_id + 1].rip) {
+            agt->rip_id++;
+            minfo("Trying next server ip in the line: '%s'.", agt->server[agt->rip_id].rip);
+            g_attempts += 5;
+            sleep(g_attempts < agt->notify_time ? g_attempts : agt->notify_time);
+        } else {
+            agt->rip_id = 0;
+            g_attempts += (5 * 3);
+            sleep(g_attempts < agt->notify_time ? g_attempts : agt->notify_time);
+        }
+    }
 
+    return;
+}
 
-            if (recv_b <= 0) {
-                /* Sleep five seconds before trying to get the reply from
-                 * the server again
-                 */
-                attempts++;
 
-                switch (recv_b) {
-                case OS_SOCKTERR:
-                    merror("Corrupt payload (exceeding size) received.");
-                    break;
-                case -1:
-                    #ifdef WIN32
+/**
+ * Holds the message reception logic for udp
+ * @param msg message to be sent
+ * @param buffer pointer to buffer where the information will be stored
+ * @param max_length size of buffer
+ * @return message_size on success
+ *         0 when all retries failed
+ * */
+static ssize_t receive_message_udp(const char *msg, char *buffer, unsigned int max_lenght) {
+    int attempts = 0;
+    ssize_t recv_b = 0;
+
+    /* Wait for server reply */
+    mwarn(AG_WAIT_SERVER, agt->server[agt->rip_id].rip);
+    sleep(1);
```

Comment:
Is there a reason to sleep here? Previously, we were sleeping only on retries.

Reply:
Yes, the first time was always failing in the UDP case. The message was sent and the recv was executed before the server had time to answer, and since it does not wait for the answer it was being considerated as a failed attempt